### PR TITLE
feat: freeze prop in Provider

### PR DESCRIPTION
**Pull request by dai-shi**
Implementation of this: https://github.com/pmndrs/jotai/issues/167#issuecomment-723515366

### CODE REVIEW

File: src/core/Provider.ts [hunk 1]
```diff
@@ -697,7 +697,9 @@ export const Provider: React.FC<{
       // XXX this is not very efficient
       mKeys(state).forEach((a) => {
         const aState = mGet(state, a) as AtomState<unknown>
-        freeze(aState.value)
```
Comment (likern) [inline src/core/Provider.ts]: If CodeSandbox CI can be package for latest PR commit that would be awesome.

Comment (dai-shi) [reply src/core/Provider.ts]: I think it did?

Comment (likern) [reply src/core/Provider.ts]: Sorry, my mistake.  CodeSandbox edited original comment. Smarter than I fought.

**Comment by likern** [comment]
I expected `freeze` will prevent my original bug saying I'm trying to modify freezed object.
```ts
const taskTags = get(taskTagsAtom);
const updatedTags = produce(Array.from(taskTags), (draft) => {
  const tagIndex = draft.findIndex((it) => it.id === action.id);
  if (tagIndex >= 0) {
    // Found tag with this id
    draft[tagIndex].isSelected = action.isSelected;
  }
});
```
But it doesn't. I don't see any errors. Still can silently update original `taskTags`

File: src/core/Provider.ts (hunk 1 above)
Comment (dai-shi) [reply src/core/Provider.ts]: You can also click the green ✔️ mark left to the commit.

**Comment by dai-shi** [comment]
Hm, that was my expectation too. 🤔

**Comment by likern** [comment]
@dai-shi I created a simple reproduction for React Native. Though I think it's not specific to it.

https://github.com/breeffy/jotai-react-native
https://expo.io/@malvikus/projects/jotai-react-native

It can be run locally using 
```bash
yarn start && yarn android
```
or 
```bash
yarn start && yarn ios
```

if `expo-cli` in installed https://docs.expo.io/.

**Comment by dai-shi** [comment]
@likern Thanks for the repro.
I tried the same in codesandbox: https://codesandbox.io/s/react-forked-hk6ei
And it doesn't work as expected.

I noticed freezing in Provider is too late in certain scenarios (derived atoms).
Let's drop this idea.

**Pull request closed without merging**